File: examples/backend-nodejs/cafebazaar-service.js
const axios = require('axios');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');

/**
 * CafeBazaar Backend Service for REST API v2 & Dynamic Pricing JWT.
 */
class CafeBazaarService {
  /**
   * @param {Object} config
   * @param {string} config.packageName - App package name (e.g. "com.example.fitme")
   * @param {string} config.apiSecretToken - CAFEBAZAAR-PISHKHAN-API-SECRET token from developer console
   * @param {string} [config.dynamicPriceSecret] - Secret key for signing Dynamic Pricing JWTs
   */
  constructor({ packageName, apiSecretToken, dynamicPriceSecret }) {
    if (!packageName || !apiSecretToken) {
      throw new Error('CafeBazaarService requires packageName and apiSecretToken');
    }
    this.packageName = packageName;
    this.apiSecretToken = apiSecretToken;
    this.dynamicPriceSecret = dynamicPriceSecret;
    this.baseUrl = 'https://pardakht.cafebazaar.ir/devapi/v2/api';

    this.http = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'CAFEBAZAAR-PISHKHAN-API-SECRET': this.apiSecretToken,
        'Content-Type': 'application/json'
      },
      timeout: 10000
    });
  }

  /**
   * Validates an In-App Purchase token against CafeBazaar servers.
   * @param {string} productId - SKU of the purchased item
   * @param {string} purchaseToken - Token returned by the client
   * @returns {Promise<{ isValid: boolean, isConsumed: boolean, isRefunded: boolean, purchaseTime: number, developerPayload: string, raw: Object }>}
   */
  async validateInAppPurchase(productId, purchaseToken) {
    try {
      const url = `/validate/${this.packageName}/inapp/${productId}/purchases/${purchaseToken}/`;
      const response = await this.http.get(url);
      const data = response.data;

      // purchaseState: 0 = OK, 1 = Refunded
      // consumptionState: 0 = Consumed, 1 = Unconsumed
      return {
        isValid: data.purchaseState === 0,
        isRefunded: data.purchaseState === 1,
        isConsumed: data.consumptionState === 0,
        purchaseTime: data.purchaseTime,
        developerPayload: data.developerPayload,
        raw: data
      };
    } catch (error) {
      if (error.response && error.response.status === 404) {
        return {
          isValid: false,
          isRefunded: false,
          isConsumed: false,
          error: error.response.data?.error || 'not_found',
          errorDescription: error.response.data?.error_description || 'Purchase not found'
        };
      }
      throw error;
    }
  }

  /**
   * Consumes an In-App Purchase token on CafeBazaar servers.
   * @param {string} purchaseToken
   * @returns {Promise<boolean>}
   */
  async consumePurchase(purchaseToken) {
    try {
      const url = `/consume/${this.packageName}/purchases/`;
      await this.http.post(url, { token: purchaseToken });
      return true;
    } catch (error) {
      if (error.response) {
        const errData = error.response.data;
        throw new Error(`Failed to consume purchase: ${errData.error} (${errData.error_description})`);
      }
      throw error;
    }
  }

  /**
   * Checks the user's active subscriptions using any recent purchase/subscription token.
   * @param {string} purchaseToken - Any purchase token from the user
   * @returns {Promise<Array<{ sku: string, autoRenewing: boolean, validUntil: Date, isExpired: boolean, linkedToken: string }>>}
   */
  async getActiveSubscriptions(purchaseToken) {
    try {
      const url = `/applications/${this.packageName}/active-subscriptions/${purchaseToken}/`;
      const response = await this.http.get(url);
      const subs = response.data.subscriptions || [];

      const now = Date.now();
      return subs.map((sub) => ({
        sku: sub.sku,
        autoRenewing: sub.autoRenewing,
        initiationTime: new Date(sub.initiationTimestampMsec),
        validUntil: new Date(sub.validUntilTimestampMsec),
        isExpired: sub.validUntilTimestampMsec <= now,
        linkedToken: sub.linkedSubscriptionToken,
        raw: sub
      }));
    } catch (error) {
      if (error.response && error.response.status === 404) {
        return []; // No active subscription found
      }
      throw error;
    }
  }

  /**
   * Checks status of a specific subscription.
   * @param {string} subscriptionSku
   * @param {string} purchaseToken
   */
  async getSubscriptionStatus(subscriptionSku, purchaseToken) {
    try {
      const url = `/applications/${this.packageName}/subscriptions/${subscriptionSku}/purchases/${purchaseToken}`;
      const response = await this.http.get(url);
      const data = response.data;
      const now = Date.now();

      return {
        autoRenewing: data.autoRenewing,
        validUntil: new Date(data.validUntilTimestampMsec),
        isExpired: data.validUntilTimestampMsec <= now,
        linkedToken: data.linkedSubscriptionToken,
        raw: data
      };
    } catch (error) {
      if (error.response && error.response.status === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Cancels subscription auto-renewal.
   * @param {string} subscriptionSku
   * @param {string} purchaseToken
   */
  async cancelSubscription(subscriptionSku, purchaseToken) {
    const url = `/applications/${this.packageName}/subscriptions/${subscriptionSku}/purchases/${purchaseToken}/cancel/`;
    const response = await this.http.post(url);
    return response.status === 200;
  }

  /**
   * Requests a refund for a purchase within 7 days.
   * @param {string} purchaseToken
   */
  async refundPurchase(purchaseToken) {
    const url = `/refund/${this.packageName}/purchases/${purchaseToken}/`;
    const response = await this.http.post(url);
    return response.data;
  }

  /**
   * Generates a signed Dynamic Pricing JWT token.
   * @param {Object} params
   * @param {number} params.priceInRials - Discounted price in Rials (must be <= registered console price)
   * @param {string} params.sku - Product SKU
   * @param {number} [params.expiresInSeconds=600] - Token expiration in seconds (default 10 mins)
   * @param {string} [params.accountId] - Optional Bazaar user account ID
   * @returns {string} Signed JWT token
   */
  generateDynamicPriceToken({ priceInRials, sku, expiresInSeconds = 600, accountId = null }) {
    if (!this.dynamicPriceSecret) {
      throw new Error('dynamicPriceSecret is required to generate dynamic pricing tokens');
    }

    const payload = {
      price: Math.floor(priceInRials),
      package_name: this.packageName,
      sku: sku,
      exp: Math.floor(Date.now() / 1000) + expiresInSeconds,
      nonce: crypto.randomBytes(16).toString('hex')
    };

    if (accountId) {
      payload.account_id = accountId;
    }

    return jwt.sign(payload, this.dynamicPriceSecret, {
      algorithm: 'HS256',
      header: {
        typ: 'JWT',
        alg: 'HS256'
      }
    });
  }
}

module.exports = CafeBazaarService;

File: examples/backend-python/cafebazaar_service.py
import time
import secrets
import base64
import requests
import jwt
from typing import Dict, Any, List, Optional
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.serialization import load_der_public_key

class CafeBazaarService:
    def __init__(self, package_name: str, api_secret_token: str, dynamic_price_secret: Optional[str] = None):
        self.package_name = package_name
        self.api_secret_token = api_secret_token
        self.dynamic_price_secret = dynamic_price_secret
        self.base_url = "https://pardakht.cafebazaar.ir/devapi/v2/api"
        self.headers = {
            "CAFEBAZAAR-PISHKHAN-API-SECRET": self.api_secret_token,
            "Content-Type": "application/json"
        }

    def validate_inapp_purchase(self, product_id: str, purchase_token: str) -> Dict[str, Any]:
        """
        Validates an in-app purchase using CafeBazaar REST API v2.
        """
        url = f"{self.base_url}/validate/{self.package_name}/inapp/{product_id}/purchases/{purchase_token}/"
        response = requests.get(url, headers=self.headers, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
            return {
                "is_valid": data.get("purchaseState") == 0,
                "is_refunded": data.get("purchaseState") == 1,
                "is_consumed": data.get("consumptionState") == 0,
                "purchase_time_ms": data.get("purchaseTime"),
                "developer_payload": data.get("developerPayload"),
                "raw": data
            }
        elif response.status_code == 404:
            return {"is_valid": False, "error": "not_found", "raw": response.json()}
        else:
            response.raise_for_status()

    def consume_purchase(self, purchase_token: str) -> bool:
        """
        Consumes a purchase token on CafeBazaar servers.
        """
        url = f"{self.base_url}/consume/{self.package_name}/purchases/"
        response = requests.post(url, headers=self.headers, json={"token": purchase_token}, timeout=10)
        return response.status_code == 200

    def get_active_subscriptions(self, token: str) -> List[Dict[str, Any]]:
        """
        Gets all active subscriptions for the given user token.
        """
        url = f"{self.base_url}/applications/{self.package_name}/active-subscriptions/{token}/"
        response = requests.get(url, headers=self.headers, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
            subs = data.get("subscriptions", [])
            now_ms = int(time.time() * 1000)
            
            result = []
            for sub in subs:
                valid_until_ms = sub.get("validUntilTimestampMsec", 0)
                result.append({
                    "sku": sub.get("sku"),
                    "auto_renewing": sub.get("autoRenewing"),
                    "valid_until_ms": valid_until_ms,
                    "is_expired": valid_until_ms <= now_ms,
                    "linked_token": sub.get("linkedSubscriptionToken"),
                    "raw": sub
                })
            return result
        elif response.status_code == 404:
            return []
        else:
            response.raise_for_status()

    def generate_dynamic_price_token(
        self,
        price_in_rials: int,
        sku: str,
        expires_in_seconds: int = 600,
        account_id: Optional[str] = None
    ) -> str:
        """
        Generates a signed JWT token for CafeBazaar Dynamic Pricing (تخفیف پویا).
        """
        if not self.dynamic_price_secret:
            raise ValueError("dynamic_price_secret is required for dynamic pricing tokens")

        payload = {
            "price": int(price_in_rials),
            "package_name": self.package_name,
            "sku": sku,
            "exp": int(time.time()) + expires_in_seconds,
            "nonce": secrets.token_hex(16)
        }

        if account_id:
            payload["account_id"] = account_id

        return jwt.encode(
            payload=payload,
            key=self.dynamic_price_secret,
            algorithm="HS256",
            headers={"typ": "JWT", "alg": "HS256"}
        )

    @staticmethod
    def verify_rsa_signature(purchase_json: str, signature_b64: str, rsa_public_key_b64: str) -> bool:
        """
        Verifies local RSA signature returned by client using Developer Public Key.
        """
        try:
            pub_key_bytes = base64.b64decode(rsa_public_key_b64)
            public_key = load_der_public_key(pub_key_bytes)
            
            sig_bytes = base64.b64decode(signature_b64)
            data_bytes = purchase_json.encode('utf-8')
            
            public_key.verify(
                sig_bytes,
                data_bytes,
                padding.PKCS1v15(),
                hashes.SHA1()
            )
            return True
        except Exception:
            return False
